tools/rewriter/DetermineAbi.cpp

```diff
@@ -261,5 +253,72 @@ FnSignature determineFnSignatureForProtoType(const clang::FunctionProtoType &fpt
       cgm,
       fpt.getCanonicalTypeUnqualified().castAs<clang::FunctionProtoType>());
 
-  return determineFnSignature(info, astContext, arch);
+  return determineAbiSignature(info, astContext, arch);
+}
+
+// Not thread safe.
+std::unordered_map<std::string, uint32_t> type_ids;
+
+uint32_t get_type_id(clang::QualType type) {
+  auto canonical_name = type.getCanonicalType().getAsString();
+  // Constructs the default value, 0, if the key doesn't exist yet.
+  auto type_id = type_ids[canonical_name];
+  if (type_id == 0) {
+    type_id = type_ids.size() + 1;
+    type_ids[canonical_name] = type_id;
+  }
+  return type_id;
+}
+
+Param determineParam(std::string name, clang::QualType type) {
+  return (Param) {
+      .name = name,
+      .type_name = type.getAsString(),
+      .canonical_type_name = type.getCanonicalType().getAsString(),
+      .type_id = get_type_id(type),
+  };
+}
```

Comment:
Nit, but I'm not a fan of `determineParam` as the name here; it doesn't seem like this function is determining a parameter, it's just obtaining the type info for a parameter.

Reply:
I named it `determine*` since all of the other functions were `determine*` like `determineAbiForDecl`, etc.  If you have a better name I can use that, but I'm not sure what to name it otherwise.

Reply:
This effectively seems to be the constructor for `Param`, so it could just be `createParam`.

Reply:
Actually I just realized #522 deletes (changes and inlines) `determineParam` so I'm just gonna leave it as is then.